internal/uvm/scsi.go

```diff
@@ -331,13 +350,147 @@ func (uvm *UtilityVM) GetScsiUvmPath(ctx context.Context, hostPath string) (stri
+		return nil, fmt.Errorf(errMsgFmt, err)
+	}
+	if err := encoder.Encode(sm.Controller); err != nil {
+		return nil, fmt.Errorf(errMsgFmt, err)
+	}
+	if err := encoder.Encode(sm.LUN); err != nil {
+		return nil, fmt.Errorf(errMsgFmt, err)
+	}
+	if err := encoder.Encode(sm.ReadOnly); err != nil {
+		return nil, fmt.Errorf(errMsgFmt, err)
+	}
+	if err := encoder.Encode(sm.AttachmentType); err != nil {
+		return nil, fmt.Errorf(errMsgFmt, err)
+	}
+	return buf.Bytes(), nil
+}
+
+// GobDecode deserializes the SCSIMount struct into the struct on which this is called
+// (i.e the sm pointer)
+func (sm *SCSIMount) GobDecode(data []byte) error {
+	buf := bytes.NewBuffer(data)
+	decoder := gob.NewDecoder(buf)
+	errMsgFmt := "failed to decode SCSIMount: %s"
+	// fields should be decoded in the same order in which they were encoded.
+	if err := decoder.Decode(&sm.HostPath); err != nil {
+		return fmt.Errorf(errMsgFmt, err)
+	}
+	if err := decoder.Decode(&sm.UVMPath); err != nil {
+		return fmt.Errorf(errMsgFmt, err)
+	}
+	if err := decoder.Decode(&sm.Controller); err != nil {
+		return fmt.Errorf(errMsgFmt, err)
+	}
+	if err := decoder.Decode(&sm.LUN); err != nil {
+		return fmt.Errorf(errMsgFmt, err)
 	}
+	if err := decoder.Decode(&sm.ReadOnly); err != nil {
+		return fmt.Errorf(errMsgFmt, err)
+	}
+	if err := decoder.Decode(&sm.AttachmentType); err != nil {
+		return fmt.Errorf(errMsgFmt, err)
+	}
+	return nil
+}
+
+// Clone function creates a clone of the SCSIMount `sm` and adds the cloned SCSIMount to
+// the uvm `vm`. If `sm` is read only then it is simply added to the `vm`. But if it is a
+// writeable mount(e.g a scratch layer) then a copy of it is made and that copy is added
+// to the `vm`.
+func (sm *SCSIMount) Clone(ctx context.Context, vm *UtilityVM, cd *cloneData) error {
+	var (
+		dstVhdPath string = sm.HostPath
+		err        error
+		dir        string
+		conStr     string = fmt.Sprintf("%d", sm.Controller)
+		lunStr     string = fmt.Sprintf("%d", sm.LUN)
+	)
+
+	if !sm.ReadOnly {
+		// Copy this scsi disk
+		// TODO(ambarve): This is a writeable SCSI mount. It can either be the
+		// scratch VHD of the UVM or it can be a SCSI mount that belongs to some
+		// container which is being automatically cloned here as a part of UVM
+		// cloning process. We will receive a request for creation of this
+		// container later on which will specify the storage path for this
+		// container.  However, that storage location is not available now so we
+		// just use the storage of the uvm instead. Find a better way for handling
+		// this. Problem with this approach is that the scratch VHD of the container
+		// will not be automatically cleaned after container exits. It will stay
+		// there as long as the UVM keeps running.
+
+		// For the scratch VHD of the VM (always attached at Controller:0, LUN:0)
+		// clone it in the scratch folder
+		dir = cd.scratchFolder
+		if sm.Controller != 0 || sm.LUN != 0 {
```

Comment:
I think we may just want to error if there are other r/w disks than the scratch specified in the template. Generally if the user has added a disk, they probably want a way to save the disk later. I think just copying the disk like this would be surprising to users since they wouldn't know where to find it.

Reply:
Good point. I think we should not allow r/w disks in the template at all. It doesn't work well with vsmb save compute system anyway. I will update this to throw proper error.

Reply:
Maybe I'm missing something, but it seems like we are still just copying any RW VHD here.

Reply:
I actually added a change[ here](https://github.com/microsoft/hcsshim/pull/840/files#diff-895c6c864fdd12684cbf2a87461df2bce4a6ee632d094c6b798b73de9baf7a7dR175-R179) to not allow any mounts into a template container at all so we should ideally never go into this if condition at all. But I will update this to just throw an error if we do.

Reply:
Actually I forgot about one use case where this is needed. The scratch VHD of the containers inside the template uvm needs to be copied and added on the cloned uvm here so we need this code. However, since we won't allow any user mounts on the template container this will never happen for any user added disks.